### PR TITLE
Implement FILEUNIT

### PR DESCRIPTION
**Pull request by andlaus**
This PR is incomplete as a full implementation would imply a major rewrite of the core parser. the rationale for this is to make it easier to use deck editors which like to add this keyword to all files (petrel?): it causes the parser not to trip if it encounters FILEUNIT and also, the `Opm::checkDeck()` function prints a warning if a unit system specified via FILEUNIT is not identical to the unit system of the deck.

### CODE REVIEW

**Comment by andlaus** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/checkDeck.cpp [hunk 1]
```diff
@@ -48,6 +48,49 @@ bool checkDeck( const Deck& deck, const Parser& parser, size_t enabledChecks) {
         deckValid = deckValid && Section::checkSectionTopology(deck, parser, ensureKeywordSection);
     }
 
+    // ensure that only a single unit system is used by the deck
+    int numUnitSystems = 0;
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/checkDeck.cpp]: I suggest an alternative implementation:
```C++
std::unordered_set<std::string> unit_systems;

if (deck.hasKeyword("METRIC"))
    unit_systems.insert("METRIC");
...
...

if (unit_systems.size() > 1) {
     parseContext.handleError(DECK_MULTIPLE_UNIT_SYSTEMS, "Multiple incompatible unit systems specified", errors);
```
}

File: src/opm/parser/eclipse/EclipseState/checkDeck.cpp [hunk 2]
```diff
@@ -48,6 +48,49 @@ bool checkDeck( const Deck& deck, const Parser& parser, size_t enabledChecks) {
...
+        ++ numUnitSystems;
+    if (deck.hasKeyword("METRIC"))
+        ++ numUnitSystems;
+    if (deck.hasKeyword("LAB"))
+        ++ numUnitSystems;
+    if (deck.hasKeyword("PVT-M"))
+        ++ numUnitSystems;
+
+    if (numUnitSystems != 1) {
+        std::string msg("Exactly one unit system must be specified by the deck (is: " + std::to_string(numUnitSystems) + ").");
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/checkDeck.cpp]: It is perfectly OK to not specify any unit system - in that case `METRIC` is the default.

File: src/opm/parser/eclipse/EclipseState/checkDeck.cpp [hunk 3]
```diff
@@ -48,6 +48,49 @@ bool checkDeck( const Deck& deck, const Parser& parser, size_t enabledChecks) {
...
+        OpmLog::warning(msg);
+        deckValid = false;
+    }
+
+    // make sure that no FILEUNIT keyword that specifies a different unit system as the
+    // deck is present
+    for (size_t kwIdx = 0; kwIdx < deck.size(); ++ kwIdx) {
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/checkDeck.cpp]: This loop can be simplified:
```C++
const std::string& unit_system = deck.getActiveUnitSystem().getName();
for (const auto& kw_ptr : deck.getKeywordList("FILEUNIT")) {
     if (kw_ptr->getRecord(0).getItem("FILE_UNIT_SYSTEM").getTrimmedString(0) != unit_system) {
            std::string msg = "Unit system f... up at " + kw_ptr->getFileName() + ":" + std::to_string(kw_ptr->getLineNumber());
             parseContext.handleError(UNIT_SYSTEM_FUCK_UP, msg , errors);  
     }
}
```

File: src/opm/parser/eclipse/EclipseState/checkDeck.cpp (hunk 2 above)
Comment (andlaus) [reply src/opm/parser/eclipse/EclipseState/checkDeck.cpp]: fixed. thanks!

**Comment by joakim-hove** [comment]
This was a new keyword to me - luckily the keyword is only used for checking also in Eclipse; i.e. the current patch is complete from a functionality point of view - although I have some suggestion for how the implementation can be simplified.

Good!

File: src/opm/parser/eclipse/EclipseState/checkDeck.cpp (hunk 3 above)
Comment (andlaus) [reply src/opm/parser/eclipse/EclipseState/checkDeck.cpp]: I picked as simpler implementation to fix this. if this turns out to be a common mistake (I doubt it), the error messages should get spiced up and something like that is needed...

**Comment by andlaus** [comment]
the keyword is quite an atrocity IMO. anyway, I've recently encountered it in the wild...

**Comment by andlaus** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
> the keyword is quite an atrocity IMO. anyway, I've recently encountered it in the wild...

Yes - I agree; but I the initial panic vexed off when I realized it was only used for *checking* in Eclipse; Eclipse does not actually handle the situation with different unit systems in different files.

**Comment by andlaus** [comment]
> the initial panic vexed off when I realized it was only used for checking in Eclipse; Eclipse does not actually handle the situation with different unit systems in different files.

ah, right. I did not see the infobox in the ECL documentation. this even means that the proposed implementation is complete? as a consequence, maybe an exception should be thrown if an error is encountered?

**Comment by joakim-hove** [comment]
>this even means that the proposed implementation is complete? 

yes - I would say so.

> as a consequence, maybe an exception should be thrown if an error is encountered?

Yes - as mentioned I would prefer if the `ParseContext` machinery was used to handle the error, then the behavior could be configurable - the default should probably be an exception.

**Comment by andlaus** [comment]
okay, I've changed the implementation to what you've suggested minus the swearing (I've been told that this is frowned upon in OPM). This implied deeper changes than before, including an API change that ripples down to opm-simulators: OPM/opm-simulators#1781.

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1781 please

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1781 please

**Comment by joakim-hove** [comment]
> okay, I've changed the implementation to what you've suggested 

Thank you!

> minus the swearing

Oh come on :-)

**Comment by andlaus** [comment]
>> minus the swearing
>
> Oh come on :-)

goddammit! I would have preferred to f*!&ing leave it in! ;)

**Comment by joakim-hove** [comment]
But the test failure is not obvious to me?

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1781 please

**Comment by andlaus** [comment]
> But the test failure is not obvious to me?

It's probably just a jenkins/github synchronization issue...

**Comment by andlaus** [comment]
yay! jenkins is happy without having changed anything here or downstream...

**Pull request merged**